### PR TITLE
Fix #1020: ext.todo todolist not linking to the page in pdflatex

### LINKED ISSUE #1020: ext.todo todolist not linking to the page in pdflatex
I activated the todo extension and use 

.. todolist::

on the first (or last) page of my document.

I created the document using "make latexpdf"

It shows the todo entries well, but the links are always to the first page of the document.

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1020
- Originally reported by: Anonymous
- Originally created at: 2012-10-19T17:33:36.546

### PR DESCRIPTION
**Pull request by Smattr**
Subject: Fix #1020: ext.todo todolist not linking to the page in pdflatex

### Feature or Bugfix

- Bugfix

### Purpose
This patch series fixes an issue where items created using the todo extension are not linked from the `todolist` summary when using the LaTeX back end. Following these changes, the "original entry" text in the resulting PDF correctly links to the todo entry to which it refers. There is also a trivial typo fix included in these changes.

### Detail
- #1020: ext.todo todolist not linking to the page in pdflatex

These changes are perhaps not as clean as they could be, so if there are any suggestions on improving them they are most welcome. I discovered some oddities while working on this in relation to the way LaTeX output is manually constructed as text blocks in this extension. For example, an exception block in this extension has the comment "ignore if no URI can be determined, e.g. for LaTeX output" but the LaTeX back end does not actually trigger the exception in question. My changes could probably do with some review from the original author of this extension, Daniel Bültmann, if he is contactable.

### Relates
N/A

### CODE REVIEW

File: sphinx/ext/todo.py [hunk 1]
```diff
@@ -212,8 +218,14 @@ def depart_todo_node(self, node):
 
 def latex_visit_todo_node(self, node):
     # type: (nodes.NodeVisitor, todo_node) -> None
-    title = node.pop(0).astext().translate(tex_escape_map)
-    self.body.append(u'\n\\begin{sphinxadmonition}{note}{%s:}' % title)
+    target = node.get('targetref')
+    title = node[0].astext().translate(tex_escape_map)
```
Comment (tk0miya) [inline sphinx/ext/todo.py]: So far, the first child node is popped out. I don't know why that is needed. But this PR removes it. Please let me know the reason.

File: sphinx/ext/todo.py [hunk 2]
```diff
@@ -212,8 +218,14 @@ def depart_todo_node(self, node):
 
 def latex_visit_todo_node(self, node):
     # type: (nodes.NodeVisitor, todo_node) -> None
-    title = node.pop(0).astext().translate(tex_escape_map)
-    self.body.append(u'\n\\begin{sphinxadmonition}{note}{%s:}' % title)
+    target = node.get('targetref')
```
Comment (tk0miya) [inline sphinx/ext/todo.py]: Could you move this before if-block?

File: tests/test_ext_todo.py [hunk 3]
```diff
@@ -84,3 +84,31 @@ def on_todo_defined(app, node):
...
+
+    # Look for the link to foo. We could equally well look for the link to bar.
+    link = r'\{\\hyperref\[\\detokenize\{(.*?foo.*?)}]\{\\sphinxcrossref{' \
+        r'\\sphinxstyleemphasis{original entry}}}}'
+    m = re.findall(link, content)
+    assert len(m) == 1
+    target = m[0]
+
+    # Look for the targets of this link.
+    labels = [m for m in re.findall(r'\label{([^}]*)}', content)
```
Comment (tk0miya) [inline tests/test_ext_todo.py]: It seems this regexp causes errors on Travis CI.

**Review by tk0miya — CHANGES_REQUESTED** [review]

File: sphinx/ext/todo.py (hunk 2 above)
Comment (Smattr) [reply sphinx/ext/todo.py]: Sorry, I don't understand this comment. As far as I can see, it is before the if-block. Maybe I'm misunderstanding which if-block you're referring to?

File: sphinx/ext/todo.py (hunk 1 above)
Comment (Smattr) [reply sphinx/ext/todo.py]: Good catch. This was unintentional. Curiously it seems to have no ill effects. I confess, I don't understand the pop logic either. `Todo.run` returns two similar nodes, which then appear to be depended on in `process_todo`. Without any further information, I've just reverted to popping the node as it was done before.

File: tests/test_ext_todo.py (hunk 3 above)
Comment (Smattr) [reply tests/test_ext_todo.py]: Naughty me didn't test on Python 3.6... It seems unknown escapes became an error after 3.5. I don't have an accessible 3.6 right now but I've pushed an optimistic attempt. If it doesn't pass I'll debug it when my 3.6 is reachable again. Apologies for the noise.

File: sphinx/ext/todo.py (hunk 2 above)
Comment (tk0miya) [reply sphinx/ext/todo.py]: Sorry, I think `target` is only related with outputting a label. So I prefer to move this line to just above of if-block. In other word, four lines bottom from here (after comment).

File: tests/test_ext_todo.py (hunk 3 above)
Comment (tk0miya) [reply tests/test_ext_todo.py]: Don't mind. We use Travis CI for this purpose :-)

**Review by tk0miya — APPROVED** [review]

**Comment by tk0miya** [comment]
LGTM with nits.

@shimizukawa could you give me advice? I can't determine this is a bug or new feature. At present, @Smattr is proposed this is a bugfix.

**Comment by Smattr** [comment]
Ah, I see what you’re asking and you’re correct it’s only needed for the
label. However, with the pop reinstated I can’t move this down because the
node collection has been modified by that point, discarding the
‘targetref’.

**Comment by Smattr** [comment]
FWIW I was not intending to introduce a new feature. This only came to my
attention because my LaTeX build kept complaining about undefined
references.

**Comment by tk0miya** [comment]
>However, with the pop reinstated I can’t move this down because the
node collection has been modified by that point, discarding the
‘targetref’.

No, they are not related operation. `node.pop(0)` removes first child of the node, and `node.get('targetref')` fetches an attribute named `targetref` from the node itself. So the pop operation does not harm another one.

**Comment by Smattr** [comment]
Yes, you're quite right. Not sure why I thought they were not orthogonal. I've swapped them as you requested.

**Comment by tk0miya** [comment]
Okay, I just talked with @shimizukawa , and we agreed this is a bug fix. So I merge this now.

**Comment by tk0miya** [comment]
Merged. Thank you for contribution!

**Comment by shimizukawa** [comment]
Thanks!

**Comment by Smattr** [comment]
Thank you!

**Pull request merged**